### PR TITLE
NES-28 Added login page and database authentication

### PR DESCRIPTION
**Pull request by benasbart**

### CODE REVIEW

File: DataAccess/_Setup/GenerateUser.py [hunk 1]
```diff
@@ -7,8 +7,12 @@ def main():
     UMAIL = os.getenv("PGADMIN_DEFAULT_EMAIL")
     UPASS = os.getenv("PGADMIN_DEFAULT_PASSWORD")
     
-    UMAIL_HASHED = bcrypt.hashpw(UMAIL.encode('utf-8'), bcrypt.gensalt())
-    UPASS_HASHED = bcrypt.hashpw(UPASS.encode('utf-8'), bcrypt.gensalt())
+    # Define a fixed salt. (random salt for now)
+    fixed_salt = b"$2b$12$VKSaJPoloZwgNhNqMJFxfu"
```
Comment (nicki419) [inline DataAccess/_Setup/GenerateUser.py:11]: Secret should not be in code. Maybe move to .env. Don't forget to put it in EXAMPLE.env too

File: Famicom/Components/Pages/Login.razor.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+using System.Text;
+using Microsoft.AspNetCore.Components;
+using DataAccess;
+using Famicom.Models;
+
+namespace Famicom.Components.Pages
```
Comment (nicki419) [inline Famicom/Components/Pages/Login.razor.cs:6]: Please move all logic to a class in Famicom.Models. have the controller (this file) only handle data binding, not actual logic. Same with the other razor.cs files

File: Famicom/Models/LoginModel.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+namespace Famicom.Models
+{
+    public class LoginModel
+    {
+        [Required]
+        [EmailAddress]
+        public string Email { get; set; } = string.Empty;
+
+        [Required]
+        public string Password { get; set; } = string.Empty;
```
Comment (nicki419) [inline Famicom/Models/LoginModel.cs:12]: Sensible data should not be public. I think hashing the username and password is the one thing we should do in the razor.cs file. So the puretext email and password never leave the class.

File: Famicom/Program.cs [hunk 4]
```diff
@@ -1,10 +1,12 @@
 using MudBlazor.Services;
 using Famicom.Components;
+using DataAccess;
 
 var builder = WebApplication.CreateBuilder(args);
 
 // Add MudBlazor services
 builder.Services.AddMudServices();
+builder.Services.AddScoped<UserRepository>();
```
Comment (nicki419) [inline Famicom/Program.cs:9]: Why do we need this here? Can this be moved to models?

File: DataAccess/UserRepository.cs [hunk 5]
```diff
@@ -211,8 +210,29 @@ public List<Cleaner> GetCleaner(string email)
...
+            {
+                cmd.Parameters.Add("@hashedEmailHex", NpgsqlTypes.NpgsqlDbType.Varchar).Value = hashedEmailHex;
+
+                using (var reader = cmd.ExecuteReader())
+                {
+                    if (reader.Read())
+                    {
+                        byte[] passwordHashBytes = (byte[])reader["upass_hash"];
+                        
+                        return BitConverter.ToString(passwordHashBytes).Replace("-", "").ToLower();
```
Comment (nicki419) [inline DataAccess/UserRepository.cs:227]: Why are we replacing characters here, and making the hash lowercase? That does not seem like a sound thing to do

File: Famicom/Components/Pages/Login.razor.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+{
+    public partial class LoginBase : ComponentBase
+    {
+        protected string? ErrorMessage { get; set; }
+
+        protected LoginModel loginModel = new LoginModel();
+
+        private readonly string fixedSalt = "$2b$12$VKSaJPoloZwgNhNqMJFxfu"; // Fixed salt for consistent hashing
```
Comment (nicki419) [inline Famicom/Components/Pages/Login.razor.cs:14]: Should be moved to .env

**Review by nicki419 — CHANGES_REQUESTED** [review]

**Review by benasbart — COMMENTED** [review]
fixed, now salt is stored encoded in base64 in the .env. (its encoded because it wasnt possible to read the salt since it has dollar signs and the dotenv library parses the dollar signs as smth else)

**Review by benasbart — COMMENTED** [review]
Moved all the logic regarding database to UserCredentialServices

File: Famicom/Components/Pages/Login.razor.cs (hunk 2 above)
Comment (benasbart) [reply Famicom/Components/Pages/Login.razor.cs:6]: Moved logic to UserCredentialsService

File: Famicom/Models/LoginModel.cs (hunk 3 above)
Comment (benasbart) [reply Famicom/Models/LoginModel.cs]: made protected

File: Famicom/Program.cs (hunk 4 above)
Comment (benasbart) [reply Famicom/Program.cs]: Removed.

File: Famicom/Components/Pages/Login.razor.cs (hunk 6 above)
Comment (benasbart) [reply Famicom/Components/Pages/Login.razor.cs]: Moved

File: Models/Services/UserCredentialsService.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    private string fixedSalt { get; set; }
+
+    private readonly UserRepository userRepository = new UserRepository();
+    public UserCredentialsService()
+    {
+        string envPath = Path.Combine(AppContext.BaseDirectory, ".env");
+        Env.Load(envPath);
+
+        string base64Salt = Env.GetString("BCRYPT_SALT");
+        fixedSalt = Encoding.UTF8.GetString(Convert.FromBase64String(base64Salt));
```
[the remark this answers is not in the document]
Comment (benasbart) [reply Models/Services/UserCredentialsService.cs:17]: Done. UserCredentialsService returns "$2b$12$VKSaJPoloZwgNhNqMJFxfu", and python script returns "b'$2b$12$VKSaJPoloZwgNhNqMJFxfu'", but they are used the same and hashed credentials match.

**Review by nicki419 — APPROVED** [review]

**Pull request merged**